main.py

```diff
@@ -0,0 +1,173 @@
+from requests.auth import HTTPBasicAuth
+from subprocess import call
+import hashlib
+import json
+import os
+import requests
+import sys
+import time
+import tinys3
+import traceback
+
+
+# Some configuration from environment variables
+STRIPE_API_KEY = os.getenv("STRIPE_API_KEY")
+REDIS_HOST = os.getenv("REDIS_HOST", "redis")
+REDIS_PORT = int(os.getenv("REDIS_PORT", "6379"))
+REDIS_RETENTION_DAYS = int(os.getenv("REDIS_RETENTION_DAYS", "30"))
+AWS_ACCESS_KEY_ID = os.getenv("AWS_ACCESS_KEY_ID")
+AWS_SECRET_ACCESS_KEY = os.getenv("AWS_SECRET_ACCESS_KEY")
+S3_ENDPOINT = os.getenv("S3_ENDPOINT")
+S3_BUCKET = os.getenv("S3_BUCKET")
+S3_PATH = os.getenv("S3_PATH")
+
+
+def archive_events():
+    print("Starting to archive events.")
+    redis = StrictRedis(host=REDIS_HOST, port=REDIS_PORT, db=0)
+    s3 = tinys3.Connection(AWS_ACCESS_KEY_ID,
+                           AWS_SECRET_ACCESS_KEY,
+                           default_bucket=S3_BUCKET,
+                           endpoint=S3_ENDPOINT,
+                           tls=True)
+
+    now = datetime.utcnow()
+    nowstring = now.strftime("%Y-%m-%d")
+
+    # 1. We collect all events objects from the query time span
+    # and group them by day.
+
+    # keep everything in RAM, keyed by day
+    events = {}
+    count = 0
+    for event in fetch_events():
+        event_datetime = datetime.fromtimestamp(event["created"])
+        datestring = event_datetime.isoformat()
+        daystring = event_datetime.strftime("%Y-%m-%d")
+        if daystring == nowstring:
+            continue
+        if daystring not in events:
+            events[daystring] = []
+        count += 1
+        events[daystring].append(event)
+    print("Fetched %d events." % count)
+
+    # 2. We go through day by day and see where new entries have been added
+    # since the last run. For these we create a backup file.
+
+    redis_retention = REDIS_RETENTION_DAYS * 24 * 60 * 60
+
+    for daystring in sorted(events.keys()):
+        print("Processing day %s with %d events" % (daystring, len(events[daystring])))
+
+        # possibly skip days already archived
+        redis_value = redis.get(daystring)
+        if redis_value is not None:
+            num_entries_before = int(redis_value)
+            if len(events[daystring]) <= num_entries_before:
+                print("Skipping day %s, already archived" % daystring)
+                continue
+
+        try:
+            dt = datetime.strptime(daystring, "%Y-%m-%d")
+
+            filename = dump_to_file(daystring, events[daystring])
+            target_path = ""
+            if S3_PATH is not None:
+                target_path = dt.strftime(S3_PATH)
+            target_path += "/" + filename
+            upload_backup(s3, filename, target_path)
+            os.remove(filename)
+
+            # write number of entries per day to redis
+            redis.setex(daystring, redis_retention, str(len(events[daystring])))
+        except Exception as ex:
```

Comment:
This catches an undefined state. You don't know what part of the backup didn't work. eg you did a successful backup but couldn't store the transient data in redis or the file couldn't be removed from disk.

I guess logs still tell you what was going wrong because of logging the exception but you could react differently depending on what kind of error happened.